src/sources/socket/udp.rs

```diff
@@ -39,41 +42,44 @@ pub fn udp(
     let out = out.sink_map_err(|e| error!("error sending event: {:?}", e));
 
     Box::new(
-        future::lazy(move || {
-            let socket = UdpSocket::bind(&address).expect("failed to bind to udp listener socket");
-
+        async move {
+            let socket = UdpSocket::bind(&address)
+                .await
+                .expect("failed to bind to udp listener socket");
             info!(message = "listening.", %address);
 
-            Ok(socket)
-        })
-        .and_then(move |socket| {
-            let host_key = host_key.clone();
-            // UDP processes messages per packet, where messages are separated by newline.
-            // And stretch to end of packet.
-            UdpFramed::with_decode(socket, BytesDelimitedCodec::new(b'\n'), true)
```

Comment:
[tokio_util::udp::UdpFramed](https://docs.rs/tokio-util/0.3.1/tokio_util/udp/struct.UdpFramed.html) do not have `with_decode` method, is it ok drop it?
In `tokio:0.1.22` [UdpFramed](https://docs.rs/tokio/0.1.22/tokio/net/struct.UdpFramed.html) also do not have this method in docs....

Reply:
Here's the relevant PR to tokio: https://github.com/tokio-rs/tokio/pull/1517/files#diff-9ceb10657fd3e9c67559c899f9bf59e7R182

It seems like the `UdpFramed` implementation was bugged, and `with_decode` is just a fix that should've been in the `new` fn.

Reply:
New impl at https://github.com/tokio-rs/tokio/blob/7e4edb89638065a041c1093b292f709d1f2190a0/tokio-util/src/udp/frame.rs#L44 looks suspicious.

Does `clear` at https://github.com/tokio-rs/tokio/blob/7e4edb89638065a041c1093b292f709d1f2190a0/tokio-util/src/udp/frame.rs#L65 means that when multiple frames arrive in a single read, all frames after the first one are dropped? This looks like a problem.

@LucioFranco, could you take a look into it if you have a minute?

Reply:
With this new `UdpFramed` `Decoder` can expect the whole payload so it can/should decode all frames. So a `Decoder` like this is required, I already wrote it in the other PR so here it is:
```

struct UdpDecoder;

impl Decoder for UdpDecoder {
    type Item = Vec<Bytes>;
    type Error = io::Error;
    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
        // UDP processes messages per payload, where messages are separated by newline
        // and stretch to end of payload.
        let mut lines = Vec::new();
        let mut decoder = BytesDelimitedCodec::new(b'\n');
        while let Some(line) = decoder.decode_eof(src)? {
            lines.push(line);
        }
        Ok(Some(lines))
    }
}
```

Reply:
An alternative is to manage the read buffer ourselves.

Reply:
Looks like pull request for this already exists https://github.com/tokio-rs/tokio/pull/1445

Reply:
Fixed in https://github.com/timberio/vector/pull/3095/commits/3cf9fb5fe6f05158b71cd5a6bc8a9128b48e9e4e
`max_length` will be fixed in another PR https://github.com/timberio/vector/issues/3117

Reply:
Yeah @MOZGIII seems like we didn't port the fix from 0.1 to 0.2 and the PR that @fanatid mentioned is the fix for always calling `decode_eof` for datagram based protocols. I just pinged the author so we could try to get that into a release.